### ISSUE TITLE
Use middleware in api route

### ISSUE THREAD

**Issue by LOLBRUHNICE** [vercel/next.js#30534]
### What version of Next.js are you using?

12.0.1

### What version of Node.js are you using?

16.12.0

### What browser are you using?

Brave

### What operating system are you using?

Windows

### How are you deploying your application?

next start

### Describe the Bug

Api route cannot get the property that assign by the middleware

### Expected Behavior

can get the property

### To Reproduce

```js
// ./pages/api/_middleware.ts
export function middleware(req: NextRequest, ev: NextFetchEvent) {
    req.something = "something"
}
```
```js
// ./pages/api/something.ts
export default function handler(
  req: NextApiRequest,
  res: NextApiResponse
) {
  console.log(req.something);
  res.send(...)
}
```

**Comment by jzyds** [comment]
mark

**Comment by lfades** [comment]
## TL;DR

The behavior you're seeing is intended, the `req` in API Routes is an instance of [http.IncomingMessage](https://nodejs.org/api/http.html#http_class_http_incomingmessage), which only exists in Node.js.

## Explanation

Middleware is expected to run in a browser environment, look at it as you're running that code in devtools, where Node.js doesn't exist but Web APIs do, that's the case for Vercel where we'll run your code in a V8 environment at the edge.

So, in Next.js Middleware the `req` is then an instance of Fetch API's [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request), which might have some similarities with Node's counterpart but they're completely different. They can't really communicate with each other by mutating the `req` object, or any other object for that matter, as they run in different environments, in different machines.

You might find this confusing if you look at the Next.js source and see that we're using Node.js to run your middlewares and "transform" the Node.js `req` into a polyfilled `Request`, but that's only because we need a way to run your middlewares locally while we find a way to actually run them in a browser, once that's done it will be much similar to how it actually works in a platform like Vercel, where we take your middleware functions and deploy them in the edge, and API routes go to serverless functions.

**Comment by isTravis** [comment]
Is there a recommended way to share data between middleware and subsequent routes (either in `pages/api` or `pages` getServerSideProps functions)?

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
